### DISCUSSION TITLE
Filter edges by timestamps

### DISCUSSION THREAD

**Discussion by YoavLotem** [pyg-team/pytorch_geometric#5196] [Q&A]
Hi team, 

I'm working on an edge classification problem with temporal constraints.
For example, given a supervision-edge from time t1 I would like to filter out all neighboring edges that came from t>t1 ("the future of t1") to prevent data leakage.  

I'm having trouble performing this type of filtering on the batch level (meaning, filtering each supervision-edge's neighboring message-passing edges using the supervision-edges' timestamps) because the message-passing edges defined by `edge_index` are not stored separately for each sample in the batch. (`edge_index`  has a `[2, num_edges]` dimensions).

**I tried the following two approaches:**

1. 1. Use a batch size of 1 and filter the entire `edge_index` by the single supervision edge's timestamp.
Worked extremely slow.

2. For each batch:
   - define multiple time-windows for supervision edges. 
   - For each time window t1-t2 (when t1<t2) : 
     -  Keep supervision-edges with timestamps between t1-t2 .
     -  Keep message-passing edges with timestamps< t1.
The loss is aggregated for each time window and the optimizer step happens at the end of each batch. 
Works faster.

**Comment by rusty1s** [comment]
We just added support for temporal-based link sampling within `LinkNeighborLoader` to master. Let me know if this helps you.

If you are batching multiple graphs, you can also leverage the `batch` vector for filtering:
```python
seed_node_time = …
mask = edge_time < seed_node_time[batch[edge_index[0]]
edge_index = edge_index[:, mask]

**Reply by YoavLotem** [reply]
Good to hear, I will check it out ASAP.
Thanks!!

**Reply by YoavLotem** [reply]
Hey, so I read the documentation.

It seems like the temporal based link sampling filters edges based on their corresponding node's timestamps. My question was regarding filtering edges based on their own timestamp.

For example, in a transaction graph (nodes are users and edges are payments), given a transaction (supervision edge) between nodes A&B that happened in time t2 (t1<t2<t3).
During message passing we would like to filter out the transaction between A&D that happened in the future (t3) of the supervision edge and keep the transaction between A&C that happened at time t1 (the past).

Does my solutions above makes sense or are their better ones currently?
Maybe its worth adding edge-wise filtering based on edge properties in the future.
Thanks in advance : )

![Picture1](https://user-images.githubusercontent.com/87936902/185079421-0ef8cd9b-f616-4ebe-9746-db422473d567.png)

**Reply by rusty1s** [reply]
Yes, this makes totally sense. Sadly, we currently support node-level time attributes right now in `NeighborLoader`. As such, you would need to manually filter out the edges :(
